compiler/damlc/lib/DA/Cli/Damlc/DependencyDb.hs

```diff
@@ -0,0 +1,204 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Cli.Damlc.DependencyDb
+    ( installDependencies
+    , dependenciesDir
+    , queryDependencies
+    , queryDependencyDalfs
+    , queryDependencyMain
+    , queryDataDependencies
+    , mainDir
+    , dalfsDir
+    , configDir
+    , sourcesDir
+    ) where
+
+import qualified "zip-archive" Codec.Archive.Zip as ZipArchive
+import Control.Exception.Safe (tryAny)
+import Control.Monad
+import DA.Daml.Compiler.Dar
+import DA.Daml.Compiler.DecodeDar (DecodedDalf(..), decodeDalf)
+import DA.Daml.Compiler.ExtractDar (ExtractedDar(..), extractDar)
+import qualified DA.Daml.LF.Ast as LF
+import DA.Daml.Options.Types
+import DA.Daml.Package.Config
+import qualified DA.Pretty
+import Data.Aeson (eitherDecodeFileStrict', encode)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BSL
+import Data.List.Extra
+import Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Text.Extended as T
+import Development.IDE.Types.Location
+import GHC.Fingerprint
+import "ghc-lib-parser" Module (unitIdString)
+import System.Directory.Extra
+import System.FilePath
+import System.IO.Extra
+
+-- Constants / Conventions
+--------------------------
+dependenciesDir :: Options -> NormalizedFilePath -> FilePath
+dependenciesDir opts projRoot =
+    fromNormalizedFilePath projRoot </> projectDependenciesDatabase </>
+    lfVersionString (optDamlLfVersion opts)
+
+fingerprintFile :: FilePath -> FilePath
+fingerprintFile depsDir = depsDir </> "fingerprint.json"
+
+mainDir, dalfsDir, configDir, sourcesDir, sdkVersionFile, depMarkerFile, dataDepMarkerFile :: FilePath -> FilePath
```

Comment:
I think it would be great to have a detailed comment somewhere here on the layout of the database. I have one main suggestion/question here at the moment:

The way this is currently setup (and that’s absolutely inherited from the previous code), you still operate on DARs. If you have two DARs that overlap in their dependencies, those dependencies will be recorded multiple times.

Can we instead flatten the layout and only store each dependency exactly once? That seems to bring a few advantages:

1. Less wasted space.
2. Each dalf can have exactly one location addressed by its package id.
3. The logic for creating the package db becomes much simpler. It only has a flat list of dalfs to deal with.

wdyt? Worth doing? I don’t mind doing that in a separate PR but I’d like to agree on whether we want to go there or not before merging this PR.

Reply:
I switched it to a flat dependency db now, but running into a problem where `daml-script-0.0.0` is a hidden package when compiling `daml-trigger`.

Reply:
sounds like we’re not properly tracking which ones are the main dalfs, i.e., which ones should be exposed?

Reply:
this is resolved now, still battling an issue with data-dependencies. The db is flat now and there is a comment describing the layout at the beginning of the file.

Reply:
data-dependency issue is resolved as well.